### PR TITLE
Wallet reorg bug fix

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

File: Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs [hunk 1]
```diff
@@ -88,44 +87,47 @@ public virtual void ProcessBlock(Block block)
...
-                    ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
-                    if (incomingBlock.Height > this.walletTip.Height)
+                    // the wallet is falling behind we need to catch up
+                    var next = this.walletTip;
+                    while(next != incomingBlock)
                     {
-                        // the wallet is falling behind we need to catch up
-                        var next = this.walletTip;
-                        while(next != incomingBlock)
+                        // while the wallet is catching up the entire node will wait
```
Comment (bokobza) [inline Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: Please use capital letters at the beginning and full stops at the end of sentences :-)

File: Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs [hunk 2]
```diff
@@ -88,44 +87,47 @@ public virtual void ProcessBlock(Block block)
...
+                        // will stop till the wallet is up to date.
+
+                        // TODO: This code should be replaced with a different approach
+                        // similar to BlockStore the wallet should be standalone and not depend on consensus
+                        // the block should be put in a queue and pushed to the wallet in an async way
+                        // if the wallet is behind it will just read blocks from store (or download in case of a pruned node)
+
+                        next = this.chain.GetBlock(next.Height +1);
+                        Block nextblock = null;
+                        while (true) //replace with cancelation token
```
Comment (bokobza) [inline Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: 'cancellation'

Comment (dangershony) [reply Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: ah well in that case I will actually replace with cancellation

**Comment by dangershony** [comment]
I think now this covers reorg, but see the comment inline about changing wallet to be stand alone.
When wallet is standalone reorg will be easier to handle (blocks will be downloaded or fetched from store).
So maybe also add a task to make the wallet standalone (like blockstore) unless we already have one.

**Comment by bokobza** [comment]
it's good to go but there is a merge conflict :anguished:

**Comment by dangershony** [comment]
fixed

**Comment by bokobza** [comment]
cool, thanks mate.

**Pull request merged**